### PR TITLE
fix(core): Observable splice index > length 

### PR DESCRIPTION
**Pull request by farfromrefug**
In javascript you can call splice with an index > length. The result is a push.
But when you do that ObservableArray index will be wrong and thus user logic will fail.
Example:
```ts
// obsarray = [0];
obsarray.splice(2, 0, 1);
// this works, the inner array is now [0,1]
// however 'change' is called with index = 2
// when the use tries to do obsarray.getItem(eventData.index) -> crash
```
We have that issue with `svelte-native` and the `spans` `ObservableArray` of `FormattedString`

### CODE REVIEW

**Review by NathanaelA — CHANGES_REQUESTED** [review]

File: packages/core/data/observable-array/index.ts [hunk 1]
```diff
@@ -246,7 +246,7 @@ export class ObservableArray<T> extends Observable {
 			eventName: CHANGE,
 			object: this,
 			action: ChangeType.Splice,
-			index: start,
+			index: Math.min(start, length),
```
[the remark this answers is not in the document]
Comment (NathanaelA) [reply packages/core/data/observable-array/index.ts:249]: Maybe it should be `Math.min(start, this._array.length-1)` so it takes in account the current length of the array when it returns the value.   The only other corner case I can think of is if you delete the last item, your index = -1...   But maybe that should be the case, since their is no valid index...

[the remark this answers is not in the document]
Comment (farfromrefug) [reply packages/core/data/observable-array/index.ts:249]: I think it is the same. Length is the size before addition 
Will look at deletion

[the remark this answers is not in the document]
Comment (NathanaelA) [reply packages/core/data/observable-array/index.ts:249]: Yes, Length is the size before the addition OR deletion.   So in addition cases the prior length would be good!    However in deletions it could very frequently be wrong. 

So instead use the `current array length - 1`  -- so that it is always a valid index except in the case you delete everything and there is only zero items...    In that case it would return -1, in that specific corner case you can't do anything with the index anyway.    So I think that is valid return in that case anyways.

So my suggestion that would fix both cases (add & delete) is 
```
index: Math.min(start, this._array.length-1);
```
Unless you have a better idea; this seems to solve both your issue that you found and the opposite issue that I just found.  :grinning:

[the remark this answers is not in the document]
Comment (farfromrefug) [reply packages/core/data/observable-array/index.ts:249]: Oh i see now. Thanks

**Review by NathanaelA — APPROVED** [review]

**Comment by maxroehrl** [comment]
@farfromrefug This MR causes a crash if `this._array.length === 0` after the splice because then we call `notify({ index: -1, ... })`.

**Comment by farfromrefug** [comment]
@maxroehrl what crash ? can you specify the example where it crashes?

**Comment by maxroehrl** [comment]
In a listview where the items are an observable array like `items = new ObersavableArray([{id: 1}]);`

Then after fetching some data I use `items.splice(0, 1, ...newData);`  to remove the placeholder item `{id: 1}`.

There is no error message but the debugger cannot step over the splice call after your change.

**Comment by farfromrefug** [comment]
@maxroehrl thank i see! Will fix it

**Comment by farfromrefug** [comment]
@maxroehrl  https://github.com/NativeScript/NativeScript/pull/8921

**Comment by NathanWalker** [comment]
@maxroehrl published in `7.0.9` - thanks for fix @farfromrefug 👍

**Comment by NathanaelA** [comment]
@maxroehrl & @farfromrefug - I found a couple other corner cases that this PR broke.   I have updated the code with PR #8968 if you both can double check that this fix doesn't break your stuff, I believe the new added tests should catch either of your prior issues.

**Comment by maxroehrl** [comment]
@NathanaelA I tested your your code and it also works for me.
